### PR TITLE
PUBDEV-5929: Added negative binomial to GLM family

### PR DESCRIPTION
**Pull request by wendycwong**
This PR completes the tasks in JIRA: https://0xdata.atlassian.net/projects/PUBDEV/issues/PUBDEV-5925?filter=myopenissues

I have completed the following:
1. Added Negative binomial to Java backend and associated calculations;
2. Added Python/R API for negative binomial.
3. Benchmark with R;
4. Benchmark with Python;
5. Java tests to 
     a. check correct gradient/hessian implementation and 
     b. compare and ensure that the model/mojo/pojo predicts are equal.

### CODE REVIEW

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 1]
```diff
@@ -479,8 +489,17 @@ public void init(boolean expensive) {
...
+          _state._ymu = new double[]{_parms._intercept ? _train.lastVec().mean() : _parms.linkInv(0)};
+/*          if (_parms._family.equals(Family.negativebinomial) && _parms._optimize_theta) {
+            double rmean = _train.lastVec().mean();
+            rmean = rmean>0?rmean:eps;
+            double rstd = _train.lastVec().sigma();
+            double temp = (rstd*rstd-rmean)/(rmean*rmean);
+            _parms._theta = temp>0?temp:eps;
+            _parms._invTheta = 1/_parms._theta;
+          }*/
+        }
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glm/GLM.java]: please ignore all parts concerning theta optimization.  I am disabling them for now but will like to add them back later.

File: h2o-algos/src/main/java/hex/glm/GLMMetricBuilder.java [hunk 2]
```diff
@@ -75,6 +76,8 @@ public double explainedDev(){
   @Override public double[] perRow(double ds[], float[] yact, double weight, double offset, Model m) {
     if(weight == 0)return ds;
     _metricBuilder.perRow(ds,yact,weight,offset,m);
+    if (_glmf._family.equals(Family.negativebinomial))
+      _sumLLH += m.likelihood(weight, yact[0], ds[0]);
     if(!ArrayUtils.hasNaNsOrInfs(ds) && !ArrayUtils.hasNaNsOrInfs(yact)) {
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glm/GLMMetricBuilder.java]: Added the likelihood calculation here to calculate the AIC.

File: h2o-algos/src/main/java/hex/glm/GLMModel.java [hunk 3]
```diff
@@ -721,10 +802,29 @@ public final double likelihood(double yr, double ym) {
...
+          d = linkInvDeriv2(x.mu)*invSum-_theta*invSum*invSum*d2*d2; // CHECKED
+          x.w = w*d;
+          x.z = eta - invSum*d2/d;
+        } else {
+          x.w = 0;
+          x.z = 0;
+        }
+      } else {
+        x.w = w / (var * d * d);  // formula did not quite work with negative binomial
+        x.z = eta + (y - x.mu) * d;
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glm/GLMModel.java]: All the likelihood/deviance, hessian and gradient calculations can be found in Negative Binomial Regression paper from NCSS.com chapter 326 pages 2 to 3.

File: h2o-algos/src/main/java/hex/glm/GLMTask.java [hunk 4]
```diff
@@ -713,14 +714,54 @@ public GLMPoissonGradientTask(Key jobKey, DataInfo dinfo, GLMParameters parms, d
...
+            double invSum =1.0/(1.0+mu*_glmf._theta);
+            double muDeriv = _glmf.linkInvDeriv(mu);
+            es[i] = ws[i] * (invSum-yr/mu+_glmf._theta*yr*invSum) * muDeriv; // gradient of -llh.  CHECKED-log/CHECKED-identity
+            l -= ws[i] * (sumOper(yr, _glmf._invTheta, 0)-(yr+_glmf._invTheta)*Math.log(1+_glmf._theta*mu)+
+                  yr*Math.log(mu)+yr*Math.log(_glmf._theta)); // store the -llh, with everything.  CHECKED-Log/CHECKED-identity
+          } else if ((mu > 0) && (yr==0)) {
+            es[i] = ws[i]*(_glmf.linkInvDeriv(mu)/(1.0+_glmf._theta*mu)); // CHECKED-log/CHECKED-identity
+            l += _glmf._invTheta*Math.log(1+_glmf._theta*mu);  //CHECKED-log/CHECKED-identity
+          } // no update otherwise
+        }
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glm/GLMTask.java]: formula can be found again from Negative Binomial Regression from NCSS.com.

File: h2o-algos/src/main/java/hex/glm/GLMTask.java [hunk 5]
```diff
@@ -1475,6 +1516,40 @@ protected void processRow(Row r) {
     }
   }
 
+  public static class GLMNegbinomialResppdate extends FrameTask2< GLMNegbinomialResppdate> {
+    private final double [] _beta; // updated  value of beta
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glm/GLMTask.java]: This one is used to calculate the response with new beta value and I hope to use it in theta optimization

File: h2o-algos/src/test/java/hex/glm/GLMBasicTestNegativebinomial.java [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+      Assert.assertTrue(model.testJavaScoring(tfr, pred, 1e-6));
+    } finally {
+      Scope.exit();
+    }
+  }
+  
+  // Test gradient/hessian/likelhood generation for negativebinomial with log link with the following paper:
+  // Negative Binomial Regression, Chapter 326, NCSS.com
+  //
+  // This test is abandoned for now until the bug in DataInfo is corrected.
```
Comment (wendycwong) [inline h2o-algos/src/test/java/hex/glm/GLMBasicTestNegativebinomial.java]: I got around this by setting response column to be the last.

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 7]
```diff
@@ -49,6 +49,7 @@
   protected boolean _cv; // flag signalling this is MB for one of the fold-models during cross-validation
   static NumberFormat lambdaFormatter = new DecimalFormat(".##E0");
   static NumberFormat devFormatter = new DecimalFormat(".##");
+  double eps = 1e-20; // smallest number to replace zero and prevent divide by zero.
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/glm/GLM.java]: there is also `hex.glm.GLMTask.EPS`

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 8]
```diff
@@ -388,11 +389,20 @@ public void init(boolean expensive) {
...
+          if (_nclass != 1) error("_family", "Poisson and Negative Binomial require the response" +
+                  " to be numeric.");
           if (_response.min() < 0)
-            error("_family", "Poisson requires response >= 0");
+            error("_family", "Poisson and Negative Binomial require response >= 0");
           if (!_response.isInt())
-            warn("_family", "Poisson expects non-negative integer response, got floats.");
+            warn("_family", "Poisson and Negative Binomial expect non-negative integer response," +
+                    " got floats.");
+          if (_parms._family.equals(Family.negativebinomial) && !_parms._optimize_theta)
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/glm/GLM.java]: please make sure _optimize_theta is not exposed to the users

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 9]
```diff
@@ -832,12 +851,58 @@ private void fitIRLSM(Solver s) {
             Log.info(LogMsg("computed in " + (t2 - t1) + "+" + (t3 - t2) + "+" + (t4 - t3) + "=" + (t4 - t1) + "ms, step = " + ls.step() + ((_lslvr != null) ? ", l1solver " + _lslvr : "")));
           } else
             Log.info(LogMsg("computed in " + (t2 - t1) + "+" + (t3 - t2) + "=" + (t3 - t1) + "ms, step = " + 1 + ((_lslvr != null) ? ", l1solver " + _lslvr : "")));
+
+/*          if (_parms._optimize_theta && (iterCnt > 1)&& _parms._family.equals(Family.negativebinomial)) {
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/glm/GLM.java]: please remove all the commented out code

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 10]
```diff
@@ -1167,6 +1232,10 @@ protected Submodel computeSubmodel(int i,double lambda) {
           else
             Log.info(LogMsg("Got " + _state.activeData().fullN() + " active columns out of " + _state._dinfo.fullN() + " total"));
           fitModel();
+          if (_parms._family.equals(Family.negativebinomial))
+            sm._trainTheta = _parms._theta;
+          if (_parms._family.equals(Family.negativebinomial))
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/glm/GLM.java]: this seems duplicate of line 1235-1236

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 11]
```diff
@@ -1238,8 +1307,18 @@ public void computeImpl() {
         if (_parms._family == Family.ordinal)
           _dinfo.addResponse(new String[]{"__glm_ExpC", "__glm_ExpNPC"}, vecs); // store eta for class C and class C-1
         else
-          _dinfo.addResponse(new String[]{"__glm_sumExp", "__glm_maxRow"}, vecs);
+          _dinfo.addResponse(new String[]{"__glm_sumExp", "__glm_maxRow"}, vecs); // will borrow one column to store estimated response for negativebinomial
+      }
+      
+      if (_parms._family.equals(Family.negativebinomial) && _parms._optimize_theta) {  // store response
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/glm/GLM.java]: remove

File: h2o-algos/src/main/java/hex/glm/GLMMetricBuilder.java [hunk 12]
```diff
@@ -27,6 +27,7 @@
   double residual_deviance;
   double null_devince;
   long _nobs;
+  double _sumLLH;
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/glm/GLMMetricBuilder.java]: follow the naming conventions, please rename to log_likelihood

File: h2o-algos/src/main/java/hex/glm/GLMModel.java [hunk 13]
```diff
@@ -183,6 +192,12 @@ public GLMModel clone2(){
     public Solver _solver = Solver.AUTO;
     public double _tweedie_variance_power;
     public double _tweedie_link_power;
+    public double _theta; // 1/k and is used by negative binomial distribution only
+    public double _invTheta;
+    public double _trainDataTheta;  // stored the theta from training data
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/glm/GLMModel.java]: will be removed according to @wendycwong

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 7 above)
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/glm/GLM.java]: Done.

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 8 above)
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/glm/GLM.java]: removed the parameter

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 1 above)
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/glm/GLM.java]: Done

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 9 above)
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/glm/GLM.java]: Done

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 10 above)
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/glm/GLM.java]: Fixed.

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 11 above)
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/glm/GLM.java]: Done

File: h2o-algos/src/main/java/hex/glm/GLMMetricBuilder.java (hunk 12 above)
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/glm/GLMMetricBuilder.java]: Done

File: h2o-algos/src/main/java/hex/glm/GLMModel.java (hunk 13 above)
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/glm/GLMModel.java]: Done

**Comment by jcrichard** [comment]
I can not make it work with grid search, the binomiale family is not recognized.
[The issue is here](https://0xdata.atlassian.net/projects/PUBDEV/issues/PUBDEV-6349?filter=allopenissues)

**Pull request merged**